### PR TITLE
feat: add initial version of Plan API implementation

### PR DESCRIPTION
**Pull request by chrisgacsal**
## Overview

* add Repository implementation
* add Service implementation
* add HTTP driver implementation

## Notes
There are a couple of missing bits which are going to be added in a separate PR:
* wire HTTP drive to the HTTP server
* use transactional database client
* add custom validation error in order to properly map status code in HTTP responses
* add tests/bug fixes

### CODE REVIEW

File: openmeter/productcatalog/plan/adapter/plan.go [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+		result = append(result, *p)
+	}
+
+	response.TotalCount = paged.TotalCount
+	response.Items = result
+
+	return response, nil
+}
+
+func (a *adapter) CreatePlan(ctx context.Context, params plan.CreatePlanInput) (*plan.Plan, error) {
```
Comment (GAlexIHU) [inline openmeter/productcatalog/plan/adapter/plan.go:112]: I suppose it's already on the way but lets use transaction utils for all repo methods

Comment (GAlexIHU) [inline openmeter/productcatalog/plan/adapter/plan.go:112]: `plan.CreatePlanInput` has Phases not CreatPhaseInput, is that intended?

File: openmeter/productcatalog/plan/service/plan.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+	// TODO(chrisgacsal): add check which makes sure that Plans with active Subscriptions are not deleted.
+
+	logger.Debug("deleting Plan")
+
+	err := s.adapter.DeletePlan(ctx, params)
+	if err != nil {
+		return fmt.Errorf("failed to delete Plan: %w", err)
+	}
+
+	logger.Debug("Plan deleted")
```
Comment (GAlexIHU) [inline openmeter/productcatalog/plan/service/plan.go:65]: just a thought, but wouldn't traces be better suited for this kind of debugging?

File: openmeter/productcatalog/plan/service/plan.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+
+	allowedPlanStatuses := []plan.PlanStatus{plan.DraftStatus, plan.ScheduledStatus}
+	planStatus := p.Status()
+	if lo.Contains(allowedPlanStatuses, p.Status()) {
+		return nil, fmt.Errorf("only Plans in %+v can be updated, but it has %s state", allowedPlanStatuses, planStatus)
+	}
+
+	logger.Debug("updating plan")
+
+	// NOTE(chrisgacsal): we only allow updating the state of the Plan via Publish/Archive
```
Comment (GAlexIHU) [inline openmeter/productcatalog/plan/service/plan.go:126]: What do you think of having separate methods for them instead?

**Review by tothandras — APPROVED** [review]

File: openmeter/productcatalog/plan/service/plan.go (hunk 2 above)
Comment (chrisgacsal) [reply openmeter/productcatalog/plan/service/plan.go:65]: Agreed. Thats my plan to do. Debug logging is there until the instrumentation is added.

File: openmeter/productcatalog/plan/service/plan.go (hunk 3 above)
Comment (chrisgacsal) [reply openmeter/productcatalog/plan/service/plan.go:126]: Hmm. I am not sure what you are referring to. We do have separate mthods for publish, archive.

**Pull request merged**